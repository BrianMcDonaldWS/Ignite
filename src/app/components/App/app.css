body {
  height: 100vh;
}

.root {
  height: 100%;
}

.contentArea {
  min-height: calc(100% - 60px - 97px);

  @media screen and (min-width: 1500px) {
    max-width: 1344px;
    width: 1344px;
    margin: auto;
  }
}

:global .blogPost {
  max-width: 700px;
  margin: auto;
  margin-bottom: 65px;
}

:global .blogHeader {
  margin-top: 2.5rem;
}

:global p.blogTitle {
  font-size: 2rem;
  font-weight: lighter;
  line-height: 2;
}

:global p.blogSubtitle {
  color: #909aa0;
  margin-bottom: 2rem;
}

:global .blogBody {
  line-height: 1.4;
  margin: 0 2rem;

  p {
    line-height: 1.9;
    margin: 15px 0;
  }

  @media screen and (min-width: 769px) {
    margin: 0 4rem;
  }
}

:global .authorImage {
  position: absolute;
  top: -30px;
  left: 50%;
  width: 60px;
  height: 60px;
  margin-left: -30px;
  border: 3px solid #ccc;
  border-radius: 50%;
}

:global .has-text-centered {
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    justify-content: center;
  }
}

h1,
h2,
h3,
h4,
h5,
h6 {
  display: flex;
  align-items: center;
  flex-wrap: wrap;

  &:not(:first-child) {
    padding-top: 25px;
  }
}

.App {
  display: flex;
  justify-content: center;

  &:global(.columns) {
    margin: 0;
  }

  h2:not(:last-child),
  h1:not(:last-child) {
    margin-bottom: 1.5rem;
  }

  :global .tag {
    margin-left: 5px;
    margin-top: 5px;
    margin-bottom: 5px;

    &:first-of-type {
      margin-left: 15px;
    }
  }
}

section {
  padding: 15px 30px;
}

a {
  color: $APP_COLOR;
}

img {
  max-width: 100%;
}

details {
  summary {
    cursor: pointer;

    &::-webkit-details-marker {
      display: none;
    }
  }

  summary:before {
    content: '\f0fe';
    font-family: Font Awesome\ 5 Free, sans-serif;
    -webkit-font-smoothing: antialiased;
    display: inline-block;
    font-style: normal;
    font-variant: normal;
    text-rendering: auto;
    line-height: 1;
    font-weight: 900;
    padding-right: 10px;
  }
}

details[open] summary:before {
  content: '\f146';
}

:global .checkbox {
  display: block;

  input[type='checkbox'],
  input[type='radio'] {
    margin-right: 10px;
  }
}

pre {
  border-left: 4px solid $APP_COLOR;

  & > code {
    display: block;
    padding: 0.5em;
  }
}

p > code {
  padding: 0.5em;
  font-size: 85%;
  line-height: 40px;
  border-radius: 3px;
  white-space: normal;
}

p {
  img[alt='caption'] {
    display: block;
    margin: auto;
  }

  img + em {
    display: block;
    text-align: center;
    margin: auto;
  }
}

:global .highlighted-line {
  display: block;
  margin: 0 -1rem;
  padding: 0 1rem;
  color: white;

  .hljs-title,
  .hljs-keyword {
    color: white;
  }
}

:global .progress-message {
  width: 100%;
  text-align: center;
}

:global .progress-with-message {
  display: flex;
  flex-direction: column;
}

:global .tile {
  h1,
  h2 {
    color: currentColor;
  }
}

:global .external-link {
  &::after {
    content: '\f14d';
    font-family: Font Awesome\ 5 Free, sans-serif;
    -webkit-font-smoothing: antialiased;
    display: inline-block;
    font-style: normal;
    font-variant: normal;
    text-rendering: auto;
    line-height: 1;
    font-weight: 900;
    padding-left: 5px;
  }
}

:global .navbar-item img {
  max-height: 3rem;
}

:global .headerLink {
  padding-left: 10px;
  opacity: 0.5;
}

:global .learnMore {
  padding: 2rem;
}

.content {
  max-width: 100%;
  margin: 0 auto 15px;
  padding-top: 0;
  padding-top: 3rem;
  padding-bottom: 5rem;

  details {
    padding-bottom: 20px;
  }

  @media screen and (min-width: 769px) {
    max-width: 950px;
  }
}

.App.blogIndex {
  margin-top: -55px;
}

.App.blogPost {
  margin-top: -115px;
}
