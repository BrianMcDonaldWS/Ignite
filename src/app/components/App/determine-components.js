import path from 'path';
import trimChar from '../../../utils/trim-char';

const getParent = (markdown, navItems, filePath) =>
  markdown.indexFiles &&
  Object.entries(markdown.indexFiles).find(([key]) => {
    const currentDir = filePath.includes('.md')
      ? path.dirname(filePath)
      : filePath;

    return (
      Object.values(navItems).includes(path.dirname(key)) &&
      trimChar(path.dirname(key), '/') === trimChar(currentDir, '/')
    );
  });

export const determineSidebar = (
  markdown,
  location,
  indexFile,
  navItems = process.env.navItems
) => {
  const filePath = location.pathname.replace('.html', '.md');
  const index = path.join(process.env.baseURL, indexFile);

  let SidebarComponent = markdown[index];
  let currentFirstPage = markdown.indexFiles[index];

  if (navItems) {
    const [parentIndex, parentPageFirstPage] =
      getParent(markdown, navItems, filePath) || [];

    if (parentIndex && parentPageFirstPage) {
      SidebarComponent = markdown[parentIndex];
      currentFirstPage = parentPageFirstPage;
    }

    if (!SidebarComponent && markdown.indexFiles) {
      const rootIndex = path.join(navItems.root, indexFile);
      SidebarComponent = markdown[rootIndex];
      currentFirstPage = markdown.indexFiles[rootIndex];
    }
  }

  window.configuration.currentFirstPage = currentFirstPage;

  return SidebarComponent;
};

export const determinePage = (
  markdown,
  location,
  indexFile,
  navItems = process.env.navItems
) => {
  const filePath = location.pathname.replace('.html', '.md');

  if (filePath.includes('blog/')) {
    return markdown[filePath];
  }

  let Page = markdown[filePath];

  if (navItems) {
    const [, parentPageFirstPage] =
      getParent(markdown, navItems, filePath) || [];

    if (parentPageFirstPage && (!Page || filePath.includes(indexFile))) {
      Page = markdown[parentPageFirstPage];
    }

    if (!Page && markdown.indexFiles) {
      const rootIndex = path.join(navItems.root, indexFile);
      Page = markdown[markdown.indexFiles[rootIndex]];
    }
  }

  const index = path.join(process.env.baseURL, indexFile);

  if ((!Page || index === filePath) && markdown.indexFiles) {
    Page = markdown[markdown.indexFiles[index]];
  }

  return Page ? Page : () => null;
};
