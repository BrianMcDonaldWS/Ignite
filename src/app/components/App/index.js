import path from 'path';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import makeClass from 'classnames';
import ReactRouterPropTypes from 'react-router-prop-types';
import scrollToElement from 'scroll-to-element';

import Header from '../Header';
import Footer from '../Footer';
import Sidebar from '../Sidebar';
import styles from './app.css';

class App extends Component {
  componentDidUpdate() {
    this.jumpToHash();
  }

  jumpToHash = () => {
    setImmediate(() => {
      const { hash } = this.props.location;
      if (hash && document.querySelector(hash)) {
        scrollToElement(hash, {
          duration: 500
        });
      } else if (!hash) {
        scrollToElement('#root', {
          duration: 1
        });
      }
    });
  };

  render() {
    const { markdown, location, index } = this.props;
    const filePath = location.pathname.substring(1);

    let Page = markdown[filePath];
    let sidebarComponent = markdown[index];

    if (process.env.navItems) {
      const parent =
        markdown.indexFiles &&
        Object.entries(markdown.indexFiles).find(([key]) => {
          return (
            Object.values(process.env.navItems).includes(path.dirname(key)) &&
            path.dirname(key) === path.dirname(filePath)
          );
        });

      if (parent) {
        sidebarComponent = markdown[parent[0]];

        console.log(filePath);
        if (!Page || filePath.includes(index)) {
          Page = markdown[parent[1]];
          console.log(markdown, parent[1], Page);
        }
      }

      if (!Page && markdown.indexFiles) {
        const rootIndex =
          process.env.navItems.root === '/'
            ? index
            : path.join(process.env.navItems.root, index);
        sidebarComponent = markdown[rootIndex];
        Page = markdown[markdown.indexFiles[rootIndex]];
      }
    }

    if (markdown.indexFiles && filePath.includes(index)) {
      Page = markdown[markdown.indexFiles[filePath]];
    }

    if (!Page && markdown.indexFiles) {
      Page = markdown[markdown.indexFiles[index]];
    }

    if (!Page) {
      Page = () => null;
    }

    return (
      <div className={styles.root}>
        <Header location={this.props.location} />

        <div id="root" className={makeClass('container', styles.contentArea)}>
          <div className={makeClass(styles.App, 'columns')}>
            <Sidebar
              className="column is-one-third-tablet is-one-quarter-desktop"
              content={sidebarComponent}
              currentPage={`${location.pathname}${
                location.hash ? location.hash : ''
              }`}
            />

            <Page
              className={makeClass(
                styles.contentWidth,
                'column',
                'content',
                'is-two-thirds-tablet',
                'is-three-quarters-desktop'
              )}
              plugins={this.props.plugins}
            />
          </div>
        </div>

        <Footer />
      </div>
    );
  }
}

App.propTypes = {
  markdown: PropTypes.object.isRequired,
  // eslint-disable-next-line react/no-typos
  location: ReactRouterPropTypes.location.isRequired,
  plugins: PropTypes.array,
  index: PropTypes.string
};

App.defaultProps = {
  plugins: [],
  index: process.env.index
};

export default App;
