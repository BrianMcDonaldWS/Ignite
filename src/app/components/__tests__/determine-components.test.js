import React from 'react';

import { determinePage, determineSidebar } from '../App/determine-components';
import renderToJson from './utils/render-to-json';

describe('determineComponents', () => {
  let markdown;

  beforeEach(() => {
    process.env.baseURL = '/';
    window.configuration = {};
    markdown = {
      '/index.md': () => <ul>List of Links</ul>,
      '/firstPage.md': () => <h1> First Page </h1>,
      indexFiles: {
        '/index.md': '/firstPage.md'
      }
    };
  });

  test('renders nothing if no pages found', () => {
    const args = [{ indexFiles: {} }, { pathname: 'notFound' }, 'index.md'];
    const Page = determinePage(...args);
    const SidebarComponent = determineSidebar(...args);

    expect(SidebarComponent).toBeUndefined();
    expect(() => renderToJson(<Page />)).not.toThrow();
  });

  test('renders first page if it exists and no pages found', () => {
    const args = [markdown, { pathname: '/notFound' }, '/index.md'];
    const Page = determinePage(...args);
    const SidebarComponent = determineSidebar(...args);

    expect(renderToJson(<SidebarComponent />)).toMatchSnapshot();
    expect(renderToJson(<Page />)).toMatchSnapshot();
  });

  test('renders matching index file if filePath matches index filename pattern', () => {
    const args = [markdown, { pathname: '/index.md' }, '/index.md'];
    const Page = determinePage(...args);
    const SidebarComponent = determineSidebar(...args);

    expect(renderToJson(<SidebarComponent />)).toMatchSnapshot();
    expect(renderToJson(<Page />)).toMatchSnapshot();
  });

  test('render matching page if found', () => {
    const navItems = {
      root: '/',
      'Name of Item': '/',
      'Name of Another Item': 'folder'
    };
    markdown['/file.md'] = () => <p> foo </p>;

    const args = [markdown, { pathname: '/file.md' }, '/index.md', navItems];
    const Page = determinePage(...args);
    const SidebarComponent = determineSidebar(...args);

    expect(renderToJson(<SidebarComponent />)).toMatchSnapshot();
    expect(renderToJson(<Page />)).toMatchSnapshot();
  });

  test('render matching index file in subpath', () => {
    const navItems = {
      root: '/',
      'Name of Item': '/',
      'Name of Another Item': '/folder'
    };
    markdown['/folder/index.md'] = () => <ul> Folder Links </ul>;
    markdown.indexFiles['/folder/index.md'] = '/folder/first.md';
    markdown['/folder/file.md'] = () => <p> foo </p>;

    const args = [
      markdown,
      { pathname: '/folder/file.md' },
      '/index.md',
      navItems
    ];
    const Page = determinePage(...args);
    const SidebarComponent = determineSidebar(...args);

    expect(renderToJson(<SidebarComponent />)).toMatchSnapshot();
    expect(renderToJson(<Page />)).toMatchSnapshot();
  });

  test('render matching index file and first page if path not found', () => {
    const navItems = {
      root: '/',
      'Name of Item': '/',
      'Name of Another Item': '/folder'
    };
    markdown['/folder/index.md'] = () => <ul> Folder Links </ul>;
    markdown.indexFiles['/folder/index.md'] = '/folder/first.md';
    markdown['/folder/first.md'] = () => <p> Folder First </p>;

    const args = [
      markdown,
      { pathname: '/folder/file.md' },
      '/index.md',
      navItems
    ];
    const Page = determinePage(...args);
    const SidebarComponent = determineSidebar(...args);

    expect(renderToJson(<SidebarComponent />)).toMatchSnapshot();
    expect(renderToJson(<Page />)).toMatchSnapshot();
  });

  test('renders root index if no matching subPath first page found', () => {
    const navItems = {
      root: '/',
      'Name of Item': '/',
      'Name of Another Item': '/folder'
    };
    markdown.indexFiles['/folder/index.md'] = '/folder/first.md';

    const args = [
      markdown,
      { pathname: '/folder/file.md' },
      '/index.md',
      navItems
    ];
    const Page = determinePage(...args);
    const SidebarComponent = determineSidebar(...args);

    expect(renderToJson(<SidebarComponent />)).toMatchSnapshot();
    expect(renderToJson(<Page />)).toMatchSnapshot();
  });

  test('render matching subpath index file and first page if path not found', () => {
    const navItems = {
      root: '/pages',
      'Name of Item': '/pages',
      'Name of Another Item': '/folder'
    };
    markdown['/pages/index.md'] = () => <ul> Pages Links </ul>;
    markdown.indexFiles['/pages/index.md'] = 'firstPage.md';
    markdown['/folder/index.md'] = () => <ul> Folder Links </ul>;
    markdown.indexFiles['/folder/index.md'] = '/folder/first.md';

    const args = [
      markdown,
      { pathname: '/folder/file.md' },
      '/index.md',
      navItems
    ];
    const Page = determinePage(...args);
    const SidebarComponent = determineSidebar(...args);

    expect(renderToJson(<SidebarComponent />)).toMatchSnapshot();
    expect(renderToJson(<Page />)).toMatchSnapshot();
  });
});
