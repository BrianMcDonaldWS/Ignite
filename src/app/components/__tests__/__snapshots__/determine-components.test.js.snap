// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`determineComponents render matching index file and first page if path not found 1`] = `
<ul>
   Folder Links 
</ul>
`;

exports[`determineComponents render matching index file and first page if path not found 2`] = `
<p>
   Folder First 
</p>
`;

exports[`determineComponents render matching index file in subpath 1`] = `
<ul>
   Folder Links 
</ul>
`;

exports[`determineComponents render matching index file in subpath 2`] = `
<p>
   foo 
</p>
`;

exports[`determineComponents render matching page if found 1`] = `
<ul>
  List of Links
</ul>
`;

exports[`determineComponents render matching page if found 2`] = `
<p>
   foo 
</p>
`;

exports[`determineComponents render matching subpath index file and first page if path not found 1`] = `
<ul>
   Folder Links 
</ul>
`;

exports[`determineComponents render matching subpath index file and first page if path not found 2`] = `
<h1>
   First Page 
</h1>
`;

exports[`determineComponents renders first page if it exists and no pages found 1`] = `
<ul>
  List of Links
</ul>
`;

exports[`determineComponents renders first page if it exists and no pages found 2`] = `
<h1>
   First Page 
</h1>
`;

exports[`determineComponents renders matching index file if filePath matches index filename pattern 1`] = `
<ul>
  List of Links
</ul>
`;

exports[`determineComponents renders matching index file if filePath matches index filename pattern 2`] = `
<h1>
   First Page 
</h1>
`;

exports[`determineComponents renders root index if no matching subPath first page found 1`] = `
<ul>
  List of Links
</ul>
`;

exports[`determineComponents renders root index if no matching subPath first page found 2`] = `
<h1>
   First Page 
</h1>
`;
